tests/byte_buf_test.c

```diff
@@ -320,42 +320,84 @@ static int s_test_buffer_printf(struct aws_allocator *allocator, void *ctx) {
     return AWS_OP_SUCCESS;
 }
 
-AWS_TEST_CASE(test_cursor_eq_case_insensitive, s_test_cursor_eq_case_insensitive)
-static int s_test_cursor_eq_case_insensitive(struct aws_allocator *allocator, void *ctx) {
+AWS_TEST_CASE(test_array_eq, s_test_array_eq)
+static int s_test_array_eq(struct aws_allocator *allocator, void *ctx) {
     (void)ctx;
     (void)allocator;
 
-    {
-        struct aws_byte_cursor a = aws_byte_cursor_from_c_str("aBc123");
-        struct aws_byte_cursor b = aws_byte_cursor_from_c_str("Abc123");
-        ASSERT_TRUE(aws_byte_cursor_eq_case_insensitive(&a, &b));
-    }
+    uint8_t a[] = {1, 2, 3};
+    uint8_t b[] = {1, 2, 3};
+    uint8_t c[] = {7, 8, 9};
+    uint8_t d[] = {1, 2, 3, 4};
 
-    {
-        struct aws_byte_cursor a = aws_byte_cursor_from_c_str("abc");
-        struct aws_byte_cursor b = aws_byte_cursor_from_c_str("xyz");
-        ASSERT_FALSE(aws_byte_cursor_eq_case_insensitive(&a, &b));
-    }
+    /* Simple */
+    ASSERT_TRUE(aws_array_eq(a, 3, b, 3));
+    ASSERT_FALSE(aws_array_eq(a, 3, c, 3));
+    ASSERT_FALSE(aws_array_eq(a, 3, d, 4));
+
+    /* Comparisons agains self */
+    ASSERT_TRUE(aws_array_eq(a, 3, a, 3));
+    ASSERT_FALSE(aws_array_eq(a, 3, a, 2));
+
+    /* Different data but size is 0 */
+    ASSERT_TRUE(aws_array_eq(a, 0, c, 0));
+
+    /* NULL inputs are OK if length is 0 */
+    ASSERT_TRUE(aws_array_eq(NULL, 0, NULL, 0));
+    ASSERT_TRUE(aws_array_eq(a, 0, NULL, 0));
+    ASSERT_TRUE(aws_array_eq(NULL, 0, b, 0));
+
+    /* Make sure we don't access NULL when length is > 0 */
+    ASSERT_FALSE(aws_array_eq(a, 3, NULL, 3));
+    ASSERT_FALSE(aws_array_eq(NULL, 3, b, 3));
+
+    /* NULL inputs are considered equivalent if their length is also the same. */
+    ASSERT_TRUE(aws_array_eq(NULL, 1, NULL, 1));
+
+    return 0;
+}
 
+AWS_TEST_CASE(test_array_eq_ignore_case, s_test_array_eq_ignore_case)
+static int s_test_array_eq_ignore_case(struct aws_allocator *allocator, void *ctx) {
+    (void)ctx;
+    (void)allocator;
     {
-        /* Vanilla strcmp() considers these equal, but not us */
-        uint8_t a_src[] = {'a', 0};
-        uint8_t b_src[] = {'a'};
-        struct aws_byte_cursor a = aws_byte_cursor_from_array(a_src, AWS_ARRAY_SIZE(a_src));
-        struct aws_byte_cursor b = aws_byte_cursor_from_array(b_src, AWS_ARRAY_SIZE(b_src));
-        ASSERT_FALSE(aws_byte_cursor_eq_case_insensitive(&a, &b));
+        uint8_t a[] = {'a', 'B', 'c', 'D', '1'};
+        uint8_t b[] = {'a', 'b', 'C', 'D', '1'}; /* same as a */
+        uint8_t c[] = {'a', 'b', 'c', 'd', '9'}; /* different */
+
+        /* Simple */
+        ASSERT_TRUE(aws_array_eq_ignore_case(a, 5, b, 5));
+        ASSERT_FALSE(aws_array_eq_ignore_case(a, 5, c, 5));
+        ASSERT_FALSE(aws_array_eq_ignore_case(a, 5, b, 3));
+
+        /* Comparisons against self */
+        ASSERT_TRUE(aws_array_eq_ignore_case(a, 5, a, 5));
+        ASSERT_FALSE(aws_array_eq_ignore_case(a, 5, a, 4));
```

Comment:
Ooh, this is an interesting case. It would be good to add an additional assert for 2 different arrays, with different lengths, just to verify that path.